GREYErrorFormatter.m

```diff
@@ -0,0 +1,195 @@
+//
+// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "GREYErrorFormatter.h"
+
+#import "GREYObjectFormatter.h"
+#import "GREYInteraction.h"
+#import "GREYErrorConstants.h"
+#import "NSError+GREYCommon.h"
+#import "GREYError+Private.h"
+#import "GREYFatalAsserts.h"
+
+#pragma mark - String Constants
+
+/**
+ * All keys that can be added to the description logger in GREYErrorFormatter.
+ */
+static NSString *const kErrorFormatterExceptionReasonKey    = @"Exception Reason";
+static NSString *const kErrorFormatterRecoverySuggestionKey = @"Recovery Suggestion";
+static NSString *const kErrorFormatterElementMatcherKey     = @"Element Matcher";
+static NSString *const kErrorFormatterSearchActionInfoKey   = @"Search Action Info";
+static NSString *const kErrorFormatterScreenshotPathsKey    = @"Screenshot Paths";
+static NSString *const kErrorFormatterUnderlyingErrorKey    = @"Underlying Error";
+static NSString *const kErrorFormatterUIHierarchyKey        = @"UI Hierarchy";
+static NSString *const kErrorFormatterStackTraceKey         = @"Stack Trace";
```

Comment:
Can we use enum and bitwise computation instead?

Reply:
That is what i initially did, but then passing around the array of the enumeration values became a larger issue than the benefit it brought. Do you know of a workaround for that?

Reply:
That would be indeed fancier, but what are the benefits? In Swift, an enum would make more sense because i could specify the enum type as the array generic element type. In Obj C, no build-time safety is enforced for the integer enum approach over the static string constant approach, right?

Reply:
It's mainly for the efficiency. To verify existence from a NSSet, you need to call string's `hash`, which is O(len(string)); OTOH bit operation is just O(1).

Is there any difference on build-time safety here? I take for either const string or enum integer, you always use it internally. Can you elaborate more?

Reply:
In Swift it'd make more sense to use an Enum because i could return an array of that enum type, but in Obj C, that is not possible. Good point on the performance difference. @tirodkar @chan-park what do you think?

Reply:
Pushed the changes.